### PR TITLE
Cleanup for in functionality

### PR DESCRIPTION
**Pull request by alschmiedt**
## The basics

- [ ] I branched from develop
- [ ] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Proposed Changes
Fixes: 
1. Going from empty workspace to stack 
2. Dealing with blocks that don't have editable field or input with connections
3. Simplifies finding the top ast node for a block 
### Test Coverage
Adds tests in mocha

Tested on:

### CODE REVIEW

File: tests/mocha/astnode_test.js [hunk 1]
```diff
@@ -316,51 +324,141 @@ suite('ASTNode', function() {
...
         assertEquals(inNode.getLocation(), outputConnection);
       });
-      test('blockToInput', function() {
+      test('fromInputToNull', function() {
+        var input = this.blocks.statementInput2.inputList[0];
+        var node = Blockly.ASTNode.createInputNode(input);
+        var inNode = node.in();
+        assertEquals(inNode, null);
+      });
+      test('fromInputToNext', function() {
```
Comment (rachel-fenichel) [inline tests/mocha/astnode_test.js]: Is this actually fromInputToPrevious?  It looks like it's going from a statement input to the previous connection on the first block in the substack.

Comment (alschmiedt) [reply tests/mocha/astnode_test.js]: Yeah it should be fromInputToPrevious. Just changed the name

File: core/ast_node.js [hunk 2]
```diff
@@ -474,6 +471,31 @@ Blockly.ASTNode.prototype.getOutAstNodeForBlock_ = function(block) {
   }
 };
 
+/**
+ * Find the first editable field or input with a connection on a given block.
+ * @param {Blockly.BlockSvg} block The source block of the current location.
```
Comment (NeilFraser) [inline core/ast_node.js]: block can't be null (block.inputList would fail), so this should be {!Blockly.BlockSvg}

File: core/ast_node.js [hunk 3]
```diff
@@ -474,6 +471,31 @@ Blockly.ASTNode.prototype.getOutAstNodeForBlock_ = function(block) {
...
+ * @return {Blockly.ASTNode} An ast node pointing to the first field or input.
+ * Null if there are no editable fields or inputs with connections on the block.
+ * @private
+ */
+Blockly.ASTNode.prototype.findFirstFieldOrInput_ = function(block) {
+  var inputs = block.inputList;
+  for (var i = 0; i < inputs.length; i++) {
+    var input = inputs[i];
```
Comment (NeilFraser) [inline core/ast_node.js]: Generally we use this form:
`  for (var i = 0, input; input = inputs[i]; i++) {`
Here and in the j loop below.

File: core/ast_node.js (hunk 2 above)
Comment (alschmiedt) [reply core/ast_node.js]: Fixed

File: core/ast_node.js (hunk 3 above)
Comment (alschmiedt) [reply core/ast_node.js]: Fixed

File: core/ast_node.js [hunk 4]
```diff
@@ -250,7 +250,7 @@ Blockly.ASTNode.prototype.findNextEditableField_ = function(location,
   var fieldRow = parentInput.fieldRow;
   var fieldIdx = fieldRow.indexOf(location);
   var startIdx = opt_first ? 0 : fieldIdx + 1;
-  for (var i = startIdx; i < fieldRow.length; i++) {
+  for (var i = startIdx, field; field = fieldRow[i]; i++) {
     var field = fieldRow[i];
```
Comment (NeilFraser) [inline core/ast_node.js]: This line is redundant.  :)

Comment (alschmiedt) [reply core/ast_node.js]: Shoot missed that. Will fix in my next PR.

**Pull request merged**